From 6270d49f7989d2b409549302c1be0c91e78a23b4 Mon Sep 17 00:00:00 2001
From: DigitalDreamtime <clive.messer@digitaldreamtime.co.uk>
Date: Sun, 8 May 2016 18:50:11 +0100
Subject: [PATCH 14/19] ES9023: i-sabre-dac-es9023-audio overlay

Add i-sabre-dac-es9023-audio overlay for Audiophonics I-Sabre DAC.

##
## Audiophonics I-Sabre DAC
##
##  card_name       = "ISabre"
##  dai_name        = "ISabre DAC"
##  dai_stream_name = "ISabre DAC HiFi"
##
# dtoverlay=i-sabre-dac-es9023-audio
# dtoverlay=i-sabre-dac-es9023-audio,bclk_ratio_int_div
# dtoverlay=i-sabre-dac-es9023-audio,bclk_ratio_int_div,384k

Signed-off-by: DigitalDreamtime <clive.messer@digitaldreamtime.co.uk>

# Conflicts:
#	arch/arm/boot/dts/overlays/Makefile
#	arch/arm/boot/dts/overlays/README
---
 arch/arm/boot/dts/overlays/Makefile                |   1 +
 arch/arm/boot/dts/overlays/README                  |  17 +++-
 .../overlays/i-sabre-dac-es9023-audio-overlay.dts  | 103 +++++++++++++++++++++
 3 files changed, 120 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/boot/dts/overlays/i-sabre-dac-es9023-audio-overlay.dts

diff --git a/arch/arm/boot/dts/overlays/Makefile b/arch/arm/boot/dts/overlays/Makefile
index 3a2352a..e8c1bd0 100644
--- a/arch/arm/boot/dts/overlays/Makefile
+++ b/arch/arm/boot/dts/overlays/Makefile
@@ -29,6 +29,7 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	hifiberry-digi-pro.dtbo \
 	hy28a.dtbo \
 	hy28b.dtbo \
+	i-sabre-dac-es9023-audio.dtbo \
 	i2c-bcm2708.dtbo \
 	i2c-gpio.dtbo \
 	i2c-mux.dtbo \
diff --git a/arch/arm/boot/dts/overlays/README b/arch/arm/boot/dts/overlays/README
index ee92426..2bc111a 100644
--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -506,13 +506,28 @@ Params: speed                   Display SPI bus speed
 
         ledgpio                 GPIO used to control backlight
 
+Name:   i-sabre-dac-es9023-audio
+Info:   Configures the Audiophonics I-Sabre DAC
+Load:   dtoverlay=i-sabre-dac-es9023-audio,<param>
+Params: 384k                    Instructs the es9023 codec driver to support
+                                352k8 and 384k sample rates.
+        bclk_ratio_int_div      Use bclk_ratio=50 for 16/24bps and
+                                bclk_ratio=100 for 32bps media when sample
+                                rate is a multiple of 8kHz and less than
+                                192kHz. Which causes the selection of the 19M2
+                                OSC as the parent for the PCM clock with an
+                                integer divider, rather than PLLD with
+                                fractional divider and MASH noise shaping.
+        card_name               Override the default, "ISabre", card name.
+        dai_name                Override the default, "ISabre DAC", dai name.
+        dai_stream_name         Override the default, "ISabre DAC HiFi",
+                                dai stream name.
 
 Name:   i2c-bcm2708
 Info:   Fall back to the i2c_bcm2708 driver for the i2c_arm bus.
 Load:   dtoverlay=i2c-bcm2708
 Params: <None>
 
-
 Name:   i2c-gpio
 Info:   Adds support for software i2c controller on gpio pins
 Load:   dtoverlay=i2c-gpio,<param>=<val>
diff --git a/arch/arm/boot/dts/overlays/i-sabre-dac-es9023-audio-overlay.dts b/arch/arm/boot/dts/overlays/i-sabre-dac-es9023-audio-overlay.dts
new file mode 100644
index 0000000..af70c9e
--- /dev/null
+++ b/arch/arm/boot/dts/overlays/i-sabre-dac-es9023-audio-overlay.dts
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2016 Clive Messer <clive.messer@digitaldreamtime.co.uk>
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPL or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Definitions for Audiophonics I-Sabre DAC using ESS Sabre ES9023 DAC
+ *  eg. dtoverlay=i-sabre-dac-es9023-audio,384k,bclk_ratio_int_div
+ */
+
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "brcm,bcm2708";
+
+	fragment@0 {
+		target = <&i2s>;
+		__overlay__ {
+			status = "okay";
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+		__overlay__ {
+			es9023_codec: es9023-codec {
+				#sound-dai-cells = <0>;
+				compatible = "ess,es9023";
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@2 {
+		target = <&sound>;
+		es9023_dac: __overlay__ {
+			compatible = "audiphonics,es9023-i-sabre-dac";
+			/*
+			 * These are the default names that will be
+			 *  applied by the es9023-dac machine driver
+			 *  based on the compatible parameter.
+			 *
+			 * es9023-dac,card-name = "ISabre";
+			 * es9023-dac,dai-name = "ISabre DAC";
+			 * es9023-dac,dai-stream-name = "ISabre DAC HiFi";
+			 */
+			status = "okay";
+			es9023-dac,cpu {
+				sound-dai = <&i2s>;
+			};
+			es9023-dac,codec {
+				sound-dai = <&es9023_codec>;
+			};
+		};
+	};
+
+	__overrides__ {
+		384k = <&es9023_codec>,"es9023,384k?";
+		bclk_ratio_int_div =
+			<&es9023_dac>,"es9023-dac,bclk-ratio-int-div?";
+		card_name = <&es9023_dac>,"es9023-dac,card-name";
+		dai_name = <&es9023_dac>,"es9023-dac,dai-name";
+		dai_stream_name = <&es9023_dac>,"es9023-dac,dai-stream-name";
+	};
+};
-- 
2.7.4

